### ISSUE TITLE
support for exposing AsyncCallback in materialized value

### ISSUE THREAD

**Issue by patriknw** [akka/akka#20503]
If callback is exposed to the outside world it can be invoked before the stage is fully initialized, resulting in error like this:

```
java.lang.IllegalStateException: not yet initialized: only setHandler is allowed in GraphStageLogic constructor
    at akka.stream.stage.GraphStageLogic.interpreter(GraphStage.scala:239)
    at akka.stream.stage.GraphStageLogic$$anon$2.invoke(GraphStage.scala:859)
    at akka.remote.artery.OutboundReplyJunction$$anon$2.sendControlMessage(Reply.scala:145)
    at akka.remote.artery.ArteryTransport.sendReply(ArteryTransport.scala:246)
    at akka.remote.artery.SystemMessageAcker$$anon$2.onPush(SystemMessageDelivery.scala:204)
    at akka.stream.impl.fusing.GraphInterpreter.processElement$1(GraphInterpreter.scala:587)
    at akka.stream.impl.fusing.GraphInterpreter.processEvent(GraphInterpreter.scala:598)
    at akka.stream.impl.fusing.GraphInterpreter.execute(GraphInterpreter.scala:539)
    at akka.stream.impl.fusing.GraphInterpreterShell.runBatch(ActorGraphInterpreter.scala:469)
    at akka.stream.impl.fusing.GraphInterpreterShell.receive(ActorGraphInterpreter.scala:408)
```

We already have the `CallbackWrapper` for exactly this purpose. It's used by `SourceQueue` and `SinkQueue`. We should probably make it public.

When doing this we should also review the implementation. It looks rather inefficient with that locking and allocations from by-name parameters.

**Comment by drewhk** [comment]
The lock is needed because there is no other way to safely implement the queue handoff in `Source.queue`. If you are interested in the detail look up the original PR with the comments by @rkuhn. I am not sure that this `CallbackWrapper` is the right approach for this general purpose. I would rather see this problem handled at the interpreter level, queueing and delaying async callbacks until stages have been properly initialized.

**Comment by eiennohito** [comment]
In addition to AsyncCallbacks, it should be possible to safely materialized ActorRef as well. I believe they should get the same treatment.

**Comment by drewhk** [comment]
They will inherit this automagically, since it is implemented on top of AsyncCallback. But you are right, ActorRefs should work, too.

**Comment by drewhk** [comment]
This is put on hold for now, the reason is that this needs some tricky coordination on the materializer side. Since we are in the process of refactoring parts of materialization I am hesitant to touch anything there (and introduce new edge cases) before we are done with that.

**Comment by fchaillou** [comment]
Hello,
I just wanted to add that even using the CallbackWrapper, there is still an issue in case of error in the stream.
Let me explain this a bit :
I'm using a copy of the `Valve` component in akka-stream-contrib that returns as its materialized value a `ValveSwitch` that allows to pause and resume a stream with its `flip` method. 

In case of an error, there is a delay before the error is propagated to my GraphStage and there is a period where i can try to invoke my asyncCallback before knowing that a failure has occured.   
In that period, calling `callback.invoke` would as usual schedule the callback to be run later but before it could be ran, then the stream will be completed by the error and so would actually never run my callback.

As a result, the callback could never complete the promise it was provided and so the caller on my materialized value will wait on an infinite future.

Here is a gist that explains the issue in code (might be easier to understand than my explanation above!)

https://gist.github.com/fchaillou/fa34d1bd46d9c0f4db7c54fc8f535fe3

Hopes, this helps !

**Comment by johanandren** [comment]
This was fixed by the invokeWithFeedback work (which buffers early invocations)
